In [23]:
products = [
    {"name":"키보드" , "price":45000 , "stock":12, "category":"전자기기"},
    {"name":"마우스" , "price":45000 , "stock":0, "category":"전자기기"},
    {"name":"모니터" , "price":45000 , "stock":4, "category":"전자기기"},
    {"name":"책상" , "price":45000 , "stock":0, "category":"가구"},
    {"name":"의자" , "price":45000 , "stock":7, "category":"가구"},
    {"name":"키보드" , "price":45000 , "stock":5, "category":"전자기기"},
    {"name":"마우스" , "price":45000 , "stock":3, "category":"전자기기"},
    {"name":"의자" , "price":45000 , "stock":3, "category":"가구"},
]
products

[{'name': '키보드', 'price': 45000, 'stock': 12, 'category': '전자기기'},
 {'name': '마우스', 'price': 45000, 'stock': 0, 'category': '전자기기'},
 {'name': '모니터', 'price': 45000, 'stock': 4, 'category': '전자기기'},
 {'name': '책상', 'price': 45000, 'stock': 0, 'category': '가구'},
 {'name': '의자', 'price': 45000, 'stock': 7, 'category': '가구'},
 {'name': '키보드', 'price': 45000, 'stock': 5, 'category': '전자기기'},
 {'name': '마우스', 'price': 45000, 'stock': 3, 'category': '전자기기'},
 {'name': '의자', 'price': 45000, 'stock': 3, 'category': '가구'}]

In [24]:
for_sales = []

found = False

for product in products:
    if product["stock"] >= 1:
        found = False
        for sale in for_sales:
            if sale["name"] == product["name"]:
                sale["stock"] += product["stock"]
                found = True
                break

        if not found:
            for_sales.append(product.copy())

for_sales

[{'name': '키보드', 'price': 45000, 'stock': 17, 'category': '전자기기'},
 {'name': '모니터', 'price': 45000, 'stock': 4, 'category': '전자기기'},
 {'name': '의자', 'price': 45000, 'stock': 10, 'category': '가구'},
 {'name': '마우스', 'price': 45000, 'stock': 3, 'category': '전자기기'}]

### 실습 1. Excel 파일 불러오기

 

해야 할 일

 

pandas를 이용해 Excel 파일을 DataFrame으로 불러옵니다.

 

먼저 코드부터 작성하기보다 인공지능에게 현재 상황과 원하는 작업을 설명해 봅니다.

In [25]:
import pandas as pd

# Excel 파일의 실제 경로
file_path = r"C:\dev\kant-axagent-study\llm-data-analysis-course\notebooks\book-text-ml\교보문고_종합_베스트셀러_상품리스트.xlsx"

# Excel 파일 불러오기
df = pd.read_excel(file_path)

# 앞의 5행 확인
df.head()

c:\dev\kant-axagent-study\.venv\Lib\site-packages\openpyxl\styles\stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


,순위,상품코드,판매상품 ID,상품명,정가,판매가,할인율,적립율,적립예정포인트,인물,출판사,발행(출시)일자,분야
0,1,9788936434120,S000000610612,소년이 온다,"15,000","13,500",10%,5%,750,한강,창비,20140519,소설
1,2,9788998441012,S000001632467,모순,"13,000","11,700",10%,5%,650,양귀자,쓰다,20130401,소설
2,3,9788997780624,S000216262757,결국 국민이 합니다,"22,000","19,800",10%,5%,"1,100",이재명,오마이북,20250415,정치/사회
3,4,9788936439743,S000216111714,혼모노,"18,000","16,200",10%,5%,900,성해나,창비,20250328,소설
4,5,9788937473401,S000200550189,급류,"14,000","12,600",10%,5%,700,정대건,민음사,20221222,소설


### 실행 결과 요약

- 데이터 불러오기: Excel 파일을 pandas로 불러와 `df`에 저장했다.
- 데이터 확인: `df.head()`를 사용해 상위 5개 상품을 확인했다.
- 분야 확인: 상위 5개 중 4개는 소설, 1개는 정치/사회 분야였다.
- 가격 확인: 상위 5개 상품은 모두 정가에서 10% 할인된 가격으로 판매되고 있었다.
- 컬럼 확인: 순위, 상품명, 가격, 할인율, 저자, 출판사, 발행일, 분야 등의 정보를 확인했다.

### 실습 2. 데이터의 크기와 컬럼 확인하기

 

Excel 파일을 불러왔으면 바로 전처리를 시작하지 않습니다.

 

먼저 어떤 데이터인지 확인합니다.

In [26]:
# 1. 전체 행과 컬럼 개수 확인
print("1. 전체 행과 컬럼 개수")
print(df.shape)

# 2. 컬럼 이름 목록 확인
print("\n2. 컬럼 이름 목록")
print(df.columns)

# 3. 각 컬럼의 자료형 확인
print("\n3. 각 컬럼의 자료형")
print(df.dtypes)

# 4. 각 컬럼의 결측치 개수 확인
print("\n4. 각 컬럼의 결측치 개수")
print(df.isna().sum())

# 5. 앞의 5행 확인
print("\n5. 앞의 5행")
print(df.head())

1. 전체 행과 컬럼 개수
(199, 13)

2. 컬럼 이름 목록
Index(['순위', '상품코드', '판매상품 ID', '상품명', '정가', '판매가', '할인율', '적립율', '적립예정포인트',
       '인물', '출판사', '발행(출시)일자', '분야'],
      dtype='str')

3. 각 컬럼의 자료형
순위          int64
상품코드        int64
판매상품 ID       str
상품명           str
정가            str
판매가           str
할인율           str
적립율           str
적립예정포인트       str
인물            str
출판사           str
발행(출시)일자    int64
분야            str
dtype: object

4. 각 컬럼의 결측치 개수
순위          0
상품코드        0
판매상품 ID     0
상품명         0
정가          0
판매가         0
할인율         0
적립율         0
적립예정포인트     0
인물          0
출판사         0
발행(출시)일자    0
분야          0
dtype: int64

5. 앞의 5행
   순위           상품코드        판매상품 ID         상품명      정가     판매가  할인율 적립율  \
0   1  9788936434120  S000000610612      소년이 온다  15,000  13,500  10%  5%   
1   2  9788998441012  S000001632467          모순  13,000  11,700  10%  5%   
2   3  9788997780624  S000216262757  결국 국민이 합니다  22,000  19,800  10%  5%   
3   4  9788936439743  S000216111714    

In [27]:
df.isna().sum().sum()

np.int64(0)

### 데이터 기본 구조 확인 결과

- 데이터 크기: 총 **199행, 13개 컬럼**으로 구성되어 있다.
- 컬럼 구성: 순위, 상품코드, 상품명, 가격, 할인율, 인물, 출판사, 발행일자, 분야 등의 정보를 포함한다.
- 결측치 확인: 모든 컬럼의 결측치가 **0개**로 확인되었다.
- 자료형 확인: `순위`, `상품코드`, `발행(출시)일자`는 `int64`, 나머지 주요 컬럼은 문자열형(`str`)이다.
- 전처리 필요: `정가`, `판매가`, `할인율`, `적립율`, `적립예정포인트`는 숫자처럼 보이지만 문자열형이라 이후 계산을 위해 형 변환이 필요할 수 있다.
- 데이터 확인: 앞의 5행을 확인한 결과 데이터가 정상적으로 불러와진 것을 확인했다. 데이터 기본 구조 확인 결과

- 데이터는 총 **199행, 13개 컬럼**으로 구성되어 있다.
- 컬럼은 순위, 상품코드, 상품명, 가격, 할인율, 인물, 출판사, 발행일자, 분야 등의 정보를 포함한다.
- 결측치는 모든 컬럼에서 **0개**로 확인되어, 현재 데이터에는 비어 있는 값이 없다.
- `순위`, `상품코드`, `발행(출시)일자`는 정수형(`int64`)이고, 나머지 주요 컬럼은 문자열형(`str`)이다.
- `정가`, `판매가`, `할인율`, `적립율`, `적립예정포인트`는 숫자처럼 보이지만 현재는 문자열형이므로, 이후 계산이나 분석을 위해 형 변환이 필요할 수 있다.
- 앞의 5행을 확인한 결과 데이터가 정상적으로 불러와진 것을 확인했다.

### 실습 3. 분석에 필요한 컬럼만 선택하기

 

원본 데이터의 모든 컬럼이 이후 분석에 필요한 것은 아닙니다.

 

이번 과정에서는 다음 컬럼을 먼저 선택합니다.

In [28]:
# 필요한 컬럼만 선택해서 새로운 DataFrame 만들기
df_books = df[
    [
        "순위",
        "판매상품 ID",
        "상품명",
        "판매가",
        "인물",
        "출판사",
        "발행(출시)일자",
        "분야"
    ]
]

# 선택된 컬럼 이름 확인
print(df_books.columns)

# 데이터가 잘 만들어졌는지 앞의 5행 확인
df_books.head()

Index(['순위', '판매상품 ID', '상품명', '판매가', '인물', '출판사', '발행(출시)일자', '분야'], dtype='str')


,순위,판매상품 ID,상품명,판매가,인물,출판사,발행(출시)일자,분야
0,1,S000000610612,소년이 온다,"13,500",한강,창비,20140519,소설
1,2,S000001632467,모순,"11,700",양귀자,쓰다,20130401,소설
2,3,S000216262757,결국 국민이 합니다,"19,800",이재명,오마이북,20250415,정치/사회
3,4,S000216111714,혼모노,"16,200",성해나,창비,20250328,소설
4,5,S000200550189,급류,"12,600",정대건,민음사,20221222,소설


### 실행 결과 요약

- 컬럼 선택: 원본 데이터에서 필요한 8개 컬럼만 선택해 `df_books`를 만들었다.
- 선택 항목: 순위, 판매상품 ID, 상품명, 판매가, 인물, 출판사, 발행(출시)일자, 분야를 선택했다.
- 결과 확인: 컬럼 목록과 앞의 5행을 확인해 원하는 데이터가 정상적으로 선택된 것을 확인했다.

### 실습 4. 컬럼 이름을 사용하기 쉽게 정리하기

 

원본 컬럼 이름은 사람이 읽기에는 좋지만 코드에서 반복해서 사용하기에는 조금 길 수 있습니다.

 

이번 실습에서는 다음처럼 간단하게 정리할 수 있습니다.

In [29]:
# 컬럼 이름 변경
df_books = df_books.rename(
    columns={
        "판매상품 ID": "판매상품ID",
        "인물": "저자",
        "발행(출시)일자": "발행일"
    }
)

# 변경된 컬럼 이름 확인
print(df_books.columns)

Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')


### 실행 결과 요약

- 컬럼명 정리: `rename()`을 사용해 일부 컬럼 이름을 더 간단하게 변경했다.
- 이름 변경: `판매상품 ID` → `판매상품ID`, `인물` → `저자`, `발행(출시)일자` → `발행일`
- 결과 확인: 변경 후 컬럼 이름이 정상적으로 적용된 것을 확인했다.

### 실습 5. 결측치 확인하기

 

결측치는 값이 비어 있는 데이터를 의미합니다.

 

데이터 분석에서는 결측치를 무조건 삭제하지 않습니다.

 

먼저 어디에 얼마나 있는지 확인하고, 분석 목적에 맞게 처리 방법을 결정합니다.

In [30]:
# 결측치가 있는 컬럼만 확인
missing = df_books.isna().sum()

print(missing[missing > 0])

Series([], dtype: int64)


### 실행 결과 요약

- 결측치 확인: `df_books`의 각 컬럼별 결측치 개수를 확인했다.
- 조건 확인: 결측치가 1개 이상인 컬럼만 따로 출력했다.
- 결과 확인: 출력 결과가 비어 있어 결측치가 있는 컬럼이 없음을 확인했다.

### 실습 6. 판매가를 숫자로 바꾸기

 

Excel에서 판매가가 다음과 같이 저장되어 있을 수 있습니다.

 

16,200

17,820

7,200

 

사람이 보기에는 숫자처럼 보이지만 pandas에서는 문자열로 읽힐 수 있습니다.

 

문자열 상태에서는 평균, 합계 등의 숫자 계산을 제대로 할 수 없습니다.

 

따라서 쉼표를 제거하고 숫자 자료형으로 변환합니다.

In [31]:
# 1. 판매가에서 쉼표 제거
df_books["판매가"] = df_books["판매가"].str.replace(",", "")

# 2. 숫자형으로 변환
df_books["판매가"] = df_books["판매가"].astype(int)

# 3. 변환 후 자료형 확인
print(df_books["판매가"].dtype)

int64


### 실행 결과 요약

- 문자열 전처리: `판매가` 컬럼에 포함된 쉼표를 제거했다.
- 자료형 변환: 문자열 형태의 `판매가`를 정수형으로 변환했다.
- 변환 확인: 자료형이 `int64`로 정상적으로 변경되었다.
- 분석 가능: 이제 `판매가`를 이용해 평균, 합계 등의 숫자 계산을 할 수 있다.

### 실습 7. 발행일을 날짜형으로 바꾸기

 

원본 데이터의 발행일은 다음처럼 보일 수 있습니다.

 

20260701

20260831

20020120

 

이 값은 연도-월-일 정보를 가지고 있지만 바로 날짜 계산에 사용하기 어려울 수 있습니다.

 

날짜 자료형으로 변환하면 연도, 월, 날짜 등을 쉽게 분석할 수 있습니다.

In [32]:
# 발행일을 날짜 자료형으로 변환
df_books["발행일"] = pd.to_datetime(
    df_books["발행일"],
    format="%Y%m%d"
)

# 변환 후 자료형 확인
print(df_books["발행일"].dtype)

# 앞의 5개 값 확인
print(df_books["발행일"].head())

datetime64[us]
0   2014-05-19
1   2013-04-01
2   2025-04-15
3   2025-03-28
4   2022-12-22
Name: 발행일, dtype: datetime64[us]


### 실행 결과 요약

- 날짜 전처리: `발행일` 컬럼을 `YYYYMMDD` 형식에서 날짜 자료형으로 변환했다.
- 자료형 변환: `발행일`의 dtype이 `datetime64[us]`로 변경되었다.
- 값 확인: `20140519` 같은 값이 `2014-05-19`처럼 읽기 쉬운 날짜 형식으로 변환되었다.
- 변환 확인: 앞의 5개 값을 확인한 결과 날짜 변환이 정상적으로 적용되었다.

### 실습 8. 문자열 앞뒤 공백 정리하기

 

상품명, 저자, 출판사, 분야는 문자열 데이터입니다.

 

문자열 앞뒤에 불필요한 공백이 있으면 같은 값도 서로 다른 데이터처럼 처리될 수 있습니다.

 

예를 들어 다음 두 값은 사람이 보기에는 같지만 문자열로는 다릅니다.

 

"소설"

"소설 "

 

이번에 정리할 컬럼

 

상품명

저자

출판사

분야


In [33]:
# 앞뒤 공백을 제거할 컬럼 목록
text_columns = ["상품명", "저자", "출판사", "분야"]

# 각 컬럼의 앞뒤 공백 제거
for col in text_columns:
    df_books[col] = df_books[col].str.strip()

# 처리 후 앞의 5개 값 확인
for col in text_columns:
    print(f"\n{col}")
    print(df_books[col].head())


상품명
0        소년이 온다
1            모순
2    결국 국민이 합니다
3           혼모노
4            급류
Name: 상품명, dtype: str

저자
0     한강
1    양귀자
2    이재명
3    성해나
4    정대건
Name: 저자, dtype: str

출판사
0      창비
1      쓰다
2    오마이북
3      창비
4     민음사
Name: 출판사, dtype: str

분야
0       소설
1       소설
2    정치/사회
3       소설
4       소설
Name: 분야, dtype: str


### 실행 결과 요약

- 문자열 전처리: `상품명`, `저자`, `출판사`, `분야` 컬럼의 앞뒤 공백을 제거했다.
- 반복 처리: 여러 문자열 컬럼을 반복문으로 한 번에 처리했다.
- 결과 확인: 각 컬럼의 앞의 5개 값을 확인한 결과 문자열이 정상적으로 정리된 것을 확인했다.

### 실습 9. 중복 데이터 확인하기

 

데이터에서 중복을 발견했다고 해서 바로 삭제하면 안 됩니다.

 

먼저 무엇을 기준으로 중복인지 생각해야 합니다.

 

예를 들어 같은 상품명이 여러 번 등장할 수 있습니다.

 

하지만 같은 제목이라도 다른 판본, 다른 상품, 개정판 등이 있을 수 있습니다.

 

따라서 이번 데이터에서는 상품명만 보고 중복 행을 바로 삭제하지 않습니다.

 

우선 상품을 구분하는 판매상품ID를 확인합니다.

In [34]:
# 1. 전체 행이 완전히 같은 중복 행 개수
print("전체 중복 행 개수:", df_books.duplicated().sum())

# 2. 판매상품ID가 중복된 행 개수
print("판매상품ID 중복 행 개수:", df_books["판매상품ID"].duplicated().sum())

# 3. 상품명이 중복된 행 개수
print("상품명 중복 행 개수:", df_books["상품명"].duplicated().sum())

전체 중복 행 개수: 0
판매상품ID 중복 행 개수: 0
상품명 중복 행 개수: 0


### 실행 결과 요약

- 전체 중복 확인: 완전히 같은 중복 행은 `0개`였다.
- 판매상품ID 중복 확인: `판매상품ID`가 중복된 행은 `0개`였다.
- 상품명 중복 확인: `상품명`이 중복된 행은 `0개`였다.
- 결과 해석: 현재 `df_books`에서는 확인한 기준으로 중복 데이터가 발견되지 않았다.

실습 10. 전처리 결과 검증하기

 

지금까지 여러 전처리를 수행했습니다.

 

마지막에는 반드시 데이터가 의도한 형태가 되었는지 다시 확인합니다.

 

확인할 내용

 

전체 행과 컬럼 개수

컬럼 이름

자료형

결측치 개수

판매가 예시

발행일 예시

중복 여부

앞의 5행

In [35]:
# 1. 전체 행과 컬럼 개수 확인
print("1. shape")
print(df_books.shape)

# 2. 컬럼 이름 확인
print("\n2. columns")
print(df_books.columns)

# 3. 각 컬럼의 자료형 확인
print("\n3. dtypes")
print(df_books.dtypes)

# 4. 각 컬럼의 결측치 개수 확인
print("\n4. 결측치 개수")
print(df_books.isna().sum())

# 5. 판매가 앞의 5개 값 확인
print("\n5. 판매가 앞의 5개 값")
print(df_books["판매가"].head())

# 6. 발행일 앞의 5개 값 확인
print("\n6. 발행일 앞의 5개 값")
print(df_books["발행일"].head())

# 7. 판매상품ID 중복 개수 확인
print("\n7. 판매상품ID 중복 개수")
print(df_books["판매상품ID"].duplicated().sum())

# 8. 전체 데이터 앞의 5행 확인
print("\n8. head()")
print(df_books.head())

1. shape
(199, 8)

2. columns
Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')

3. dtypes
순위                 int64
판매상품ID               str
상품명                  str
판매가                int64
저자                   str
출판사                  str
발행일       datetime64[us]
분야                   str
dtype: object

4. 결측치 개수
순위        0
판매상품ID    0
상품명       0
판매가       0
저자        0
출판사       0
발행일       0
분야        0
dtype: int64

5. 판매가 앞의 5개 값
0    13500
1    11700
2    19800
3    16200
4    12600
Name: 판매가, dtype: int64

6. 발행일 앞의 5개 값
0   2014-05-19
1   2013-04-01
2   2025-04-15
3   2025-03-28
4   2022-12-22
Name: 발행일, dtype: datetime64[us]

7. 판매상품ID 중복 개수
0

8. head()
   순위         판매상품ID         상품명    판매가   저자   출판사        발행일     분야
0   1  S000000610612      소년이 온다  13500   한강    창비 2014-05-19     소설
1   2  S000001632467          모순  11700  양귀자    쓰다 2013-04-01     소설
2   3  S000216262757  결국 국민이 합니다  19800  이재명  오마이북 2025-04-15  정치/사회
3   4  S000216111714   

### 최종 검증 결과 요약

- 데이터 크기 확인: 전처리된 `df_books`는 **199행, 8개 컬럼**으로 구성되어 있다.
- 컬럼 구성 확인: 순위, 판매상품ID, 상품명, 판매가, 저자, 출판사, 발행일, 분야 컬럼으로 정리되어 있다.
- 자료형 확인: `판매가`는 `int64`, `발행일`은 `datetime64[us]`로 변환되어 있다.
- 결측치 확인: 확인된 컬럼에서 결측치는 모두 `0개`였다.
- 값 확인: 판매가는 숫자형 값으로, 발행일은 날짜 형식으로 표시되고 있다.
- 데이터 확인: 앞의 5행을 통해 전처리된 데이터가 원하는 형태로 구성된 것을 확인했다.

### 실습 11. 전처리 데이터 저장하기

 

다음 Chapter에서는 이 데이터를 다시 사용합니다.

 

따라서 전처리한 DataFrame을 CSV 파일로 저장합니다.

 

추천 파일명은 다음과 같습니다.

 

book_bestseller_clean.csv

 

한글 데이터가 Excel에서 깨지는 문제를 줄이기 위해 utf-8-sig 인코딩을 사용할 수 있습니다.

In [36]:
df_books.to_csv("book_bestseller_clean.csv", encoding="utf-8-sig", index=False)

### 실행 결과 요약

- CSV 저장: 전처리가 끝난 `df_books`를 `book_bestseller_clean.csv` 파일로 저장했다.
- 인코딩 설정: `utf-8-sig`를 사용해 Windows와 Excel에서도 한글이 잘 보이도록 했다.
- 인덱스 제외: `index=False`를 사용해 DataFrame의 인덱스는 저장하지 않았다.
- 결과 확인: 현재 Notebook과 같은 폴더에 CSV 파일이 정상적으로 생성된 것을 확인했다.

### 12. 인공지능을 이용해 결과 Markdown 작성하기

 

실습을 완료하면 Notebook에 이번 작업을 설명하는 Markdown을 작성합니다.

 

여기서도 인공지능을 사용할 수 있습니다.

 

단, 인공지능이 실행 결과를 추측하게 하면 안 됩니다.

 

Notebook에서 실제로 확인한 숫자를 프롬프트에 넣어야 합니다.

### 전처리 결과 정리

#### 1. 무엇을 확인했는지
- 원본 데이터는 **199행, 13개 컬럼**이었다.
- 필요한 **8개 컬럼**만 선택했다.
- `저자`와 `분야`의 결측치는 각각 **0개**였다.
- `판매상품ID` 중복은 **0개**였다.

#### 2. 어떤 전처리를 했는지
- 필요한 컬럼만 선택하고 컬럼 이름을 정리했다.
- 문자열 컬럼의 앞뒤 공백을 제거했다.
- `판매가`를 숫자형인 `int64`로 변환했다.
- `발행일`을 날짜형인 `datetime64[us]`로 변환했다.

#### 3. 다음 분석을 위한 상태
- 결측치와 중복 여부를 확인했고, 가격과 날짜도 분석에 사용할 수 있는 자료형으로 정리했다.
- 따라서 현재 데이터는 가격, 출판사, 분야, 발행일 등을 기준으로 기본 분석을 진행할 수 있는 상태이다.

### 14. 간단한 확인 문제

 

Notebook의 마지막에 다음 질문에 본인의 답을 Markdown으로 작성해 봅니다.

 

질문 1

 

왜 모든 원본 컬럼을 사용하지 않고 필요한 컬럼만 선택했나요? 
불필요한 정보 제거
 

질문 2

 

판매가가 문자열이라면 어떤 문제가 발생할 수 있나요? 
합계 계산시 문제 발생 가능성 있음
 

질문 3

 

결측치를 확인한 뒤 바로 모든 행을 삭제하면 안 되는 이유는 무엇인가요? 어떤 이유로 결측치가 발생했을지 모르기 때문

 

질문 4

 

같은 상품명이 두 번 등장했다고 해서 바로 중복 데이터라고 판단하기 어려운 이유는 무엇인가요?
제목은 같아도 저자, 출판사, 개정판, 특별판 등 여러 가능성이 있기 때문 

질문 5

 

인공지능이 작성한 코드가 오류 없이 실행되었다면 그것만으로 분석이 끝난 것일까요?
직접 비교 확인 해야함